### DISCUSSION TITLE
I2S record and play at the same time

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#9243] [Libraries & Drivers] [answered]
Hi,
Is it possible to initiate an **audio_in** and an **audio_out** at the same time to record and play some audio at (quasi) the same time (non-blocking or using the uasyncio way)?

I would like to use a Teensy audio board with an ESP32. (The SGTL5000 has 1 I2s interface SCK, WS, D_IN and D_OUT)

Can I just use this? Or do I need 2 physical I2S busses? I base my experiments on @miketeachman's great implementation and examples at: [https://github.com/miketeachman/micropython-i2s-examples](https://github.com/miketeachman/micropython-i2s-examples)

```
from machine import I2S
from machine import Pin

sck_pin = Pin(14)   # Serial clock output
ws_pin = Pin(13)    # Word clock output
sd_out_pin = Pin(12)    # Serial data output
sd_in_pin = Pin(15)    # Serial data input

audio_out = I2S(0,
                sck=sck_pin, ws=ws_pin, sd=sd_out_pin,
                mode=I2S.TX,
                bits=16,
                format=I2S.MONO,
                rate=22050,
                ibuf=20000)

audio_in = I2S(0,
               sck=sck_pin, ws=ws_pin, sd=sd_in_pin,
               mode=I2S.RX,
               bits=16,
               format=I2S.MONO,
               rate=22050,
               ibuf=20000)
```

**Comment by miketeachman** [comment, marked as answer]
Hi,
The MicroPython I2S implementation is half-duplex so it will not be possible to use one I2S bus for both read and write operations.  Unfortunately, using 2 I2S busses will not work either as each ESP32 I2S peripheral outputs SCK and WS signals that are asynchronous.  

The following is a technical explanation on why two busses will not work.  

After configuring two I2S busses, one TX and the other RX, an oscilloscope shows that the SCK and WS signals for TX and RX do not line up.  This situation causes a problem when using the SGTL5000 codec, which only accepts one SCK and one WS signal.  Suppose the SCK and WS signals for the I2S TX bus are connected to the SCK and WS signals on the SGTL5000 codec (and the SCK and WS signals for the I2S RX bus are not used).  The codec will "clock in" the TX data using the SCK and WS signals from the I2S TX bus.  This results in the codec working for the TX direction.  The expected audio sounds will be created.  But, for the RX direction there is a problem -- the codec will send RX data to the 2nd I2S bus using the SCK and WS signals generated from the I2S TX bus.  This is where the problem appears - for correct operation the I2S RX bus needs the first bit of each 32-bit audio sample aligned with its own SCK and WS signals. But,  the I2S RX bus receives each 32-bit audio sample that has been aligned with the SCK and WS TX signals (which are not aligned with the SCK and WS RX signals).  So, the I2S RX bus clocks in each audio sample in the wrong bit location.  e.g.  the 5th sample bit might be clocked in as the 0th sample bit.  The result will be noise received on the RX bus.

A possible solution is using two Teensy audio boards, one for each of the TX and RX busses.

One other challenge using the ESP32 with the SGTL5000 codec is lack of a Master clock.  The SGTL5000 codec requires a Master clock which is not provided by the MicroPython I2S implementation for the ESP32.  Using the `machine.PWM` class to create the Master clock frequency is a possible workaround.  I have not tried this workaround.

I hope this helps!

Mike

**Reply by saraverbeecke** [reply]
Ok @miketeachman that is clear to me now, thank you for your reply!
Assume I replace the SGTL5000 with one ADC (like the [PCM1808](https://www.ti.com/lit/ds/symlink/pcm1808.pdf?ts=1662644732228&ref_url=https%253A%252F%252Fwww.google.com%252F)) and one DAC (like the [PCM1780](https://www.ti.com/lit/ds/symlink/pcm1780.pdf?ts=1662655682916&ref_url=https%253A%252F%252Fwww.google.com%252F))

Should I use slave or master mode? Should generating a mck with a PWM be reliable or what do you think is best using the 2 I2S busses on the ESP32? (Sorry but I'm new to I2S... Do I understand correctly that using the ADC in master mode, it generates its own clock, in that case is the ESP I2S bus working independently, meaning filling the DMA buffer without problems?)

**Reply by miketeachman** [reply]
I suggest using one of the simple I2S devices that only need SCK, WS, and SD signals (no Master clock), such as:

- INMP441 (ADC)
- MAX98357A (DAC)
- UDA1334A (DAC)
- PCM5102 (DAC)

All of these devices work in [peripheral mode](https://blog.adafruit.com/2021/06/24/controller-and-peripheral/)

There are many breakout boards that use these parts, available at Adafruit, Aliexpress, etc.

**Reply by saraverbeecke** [reply]
@miketeachman Thanks for the answer with the suggestions!

**Comment by xlla** [comment]
can we amend the i2s implement code to support create a full-duplex i2s class?
the upstream espressif document says "Currently, STD and TDM communication modes supports full-duplex mode in the following way"

**Comment by A622266** [comment]
Is full-duplex support for i2s available on any of the platforms yet?  I'd like to use I2S with an RP2040, but if there was something else supported maybe I could change to that ..

**Reply by Josverl** [reply]
> Is full-duplex support for i2s available on any of the platforms yet?

No, `machine.I2S()` only [supports `I2S.TX` and `.RX`](https://docs.micropython.org/en/latest/library/machine.I2S.html#constants)

I is possible to use 2 busses on several ports including the rp2040, however:
- _"The I2S class supports controller operation. Peripheral operation is not supported."_  So you cannot configure one I2S block to act as the clock master and the other as the clock slave. This may/will eventually cause buffer overruns/underruns when used as passthrough ( RX --> Processing --> TX ) 
- asyncio : if an other asyncio task takes too much time / does not yield soon enough - a channel will likely starve / overflow - you will notice this in the audio
